File: leetcode/0350-intersection-of-two-arrays-II/solution.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        """
        RUNTIME: 52 ms (90.62%)
        MEMORY: 14 MB (52.91%)
        """
        num_dict1, num_dict2 = {}, {}

        if len(num_dict1) > len(num_dict2):
            num_dict1, num_dict2 = num_dict2, num_dict1

        for i in nums1:
            if i in num_dict1:
                num_dict1[i] += 1
            else:
                num_dict1[i] = 1
        for i in nums2:
            if i in num_dict1:
                if i in num_dict2:
                    num_dict2[i] += 1
                else:
                    num_dict2[i] = 1

        intersected = []
        for each in num_dict2:
            intersected += [each] * min(num_dict1[each], num_dict2[each])
        return intersected

File: leetcode/0350-intersection-of-two-arrays-II/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 6 ms (85.10%)
     * MEMORY: 10.8 MB (21.13%)
    */
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int, int> hashmap1, hashmap2;
        vector<int> intersected;

        if (nums1.size() > nums2.size()) {
            vector<int> tmp = nums2;
            nums2 = nums1;
            nums1 = tmp;
        }
        for (int i : nums1) {
            ++hashmap1[i];
        }
        for (int i : nums2) {
            if (hashmap1.find(i) != hashmap1.end()) {
                ++hashmap2[i];
            }
        }

        for (auto i : hashmap2) {
            int tmp = i.first;
            int repeat = hashmap1[tmp] > hashmap2[tmp] ? hashmap2[tmp] : hashmap1[tmp]; 
            for (int j = 0; j < repeat; j++) {
                intersected.push_back(i.first);
            }
        }
        return intersected;
    }
};
